csharp/src/Drivers/BigQuery/BigQueryStatement.cs

```diff
@@ -0,0 +1,394 @@
+        }
+
+        public IReadOnlyDictionary<string, string>? Options { get; set; }
+
+        public override QueryResult ExecuteQuery()
+        {
+            QueryOptions? queryOptions = ValidateOptions();
+            BigQueryJob job = this.client.CreateQueryJob(SqlQuery, null, queryOptions);
+            BigQueryResults results = job.GetQueryResults();
+
+            BigQueryReadClientBuilder readClientBuilder = new BigQueryReadClientBuilder();
+            readClientBuilder.Credential = this.credential;
+            BigQueryReadClient readClient = readClientBuilder.Build();
+
+            // TODO: translate the schema
+
+            string table = $"projects/{results.TableReference.ProjectId}/datasets/{results.TableReference.DatasetId}/tables/{results.TableReference.TableId}";
+
+            ReadSession rs = new ReadSession { Table = table, DataFormat = DataFormat.Arrow };
+            ReadSession rrs = readClient.CreateReadSession("projects/" + results.TableReference.ProjectId, rs, 1);
+
+            long totalRows = results.TotalRows == null ? -1L : (long)results.TotalRows.Value;
+            IArrowArrayStream stream = new MultiArrowReader(TranslateSchema(results.Schema), rrs.Streams.Select(s => ReadChunk(readClient, s.Name)));
+
+            return new QueryResult(totalRows, stream);
+        }
+
+        public override UpdateResult ExecuteUpdate()
+        {
+            BigQueryResults result = this.client.ExecuteQuery(SqlQuery, parameters: null);
+            long updatedRows = result.NumDmlAffectedRows == null ? -1L : result.NumDmlAffectedRows.Value;
+
+            return new UpdateResult(updatedRows);
+        }
+
+        public override object GetValue(IArrowArray arrowArray, Field field, int index)
+        {
+            if(arrowArray == null) throw new ArgumentNullException(nameof(arrowArray));
+            if (field == null) throw new ArgumentNullException(nameof(field));
+            if (index < 0) throw new ArgumentOutOfRangeException(nameof(index));
+
+            try
+            {
+                switch (arrowArray)
+                {
+                    case Int64Array int64Array:
+                        return int64Array.GetValue(index);
+                    case DoubleArray doubleArray:
+                        return doubleArray.GetValue(index);
+                    case Decimal128Array decimal128Array:
+                        return decimal128Array.GetValue(index);
+                    case Decimal256Array decimal256Array:
+                        return decimal256Array.GetValue(index);
+                    case BooleanArray booleanArray:
+                        return booleanArray.GetValue(index);
+                    case StringArray stringArray:
+                        return stringArray.GetString(index);
+                    case BinaryArray binaryArray:
+
+                        ReadOnlySpan<byte> bytes = binaryArray.GetBytes(index);
+
+                        if (bytes != null)
```

Comment:
I'm not sure this will work as expected because ReadOnlySpan is a ref struct. My expectation is that this would either never evaluate to true or would evaluate to true both for an empty binary value and for null even though these are not semantically the same. It would be good to have test data which reflects this i.e. for both string and binary columns, have a null value, an empty value and non-empty value and see that we can distinguish them.

Reply:
StringArray.GetString has a related problem. This was fixed in Arrow five days ago in GH-37861, but that fix did not make it into the 14.0 release branch. Will the current C# code work against the 14.0 release branch or is it dependent on exposing the Concatenator functionality which didn't make the release?

Reply:
Okay, then only the GetBytes would need to be fixed.